flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcTableSource.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.orc;
+
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.DataSet;
+import org.apache.flink.api.java.ExecutionEnvironment;
+import org.apache.flink.api.java.typeutils.RowTypeInfo;
+import org.apache.flink.table.sources.BatchTableSource;
+import org.apache.flink.table.sources.ProjectableTableSource;
+import org.apache.flink.table.sources.TableSource;
+import org.apache.flink.types.Row;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.orc.TypeDescription;
+
+/**
+ * Creates a TableSource to read ORC file.
+ *
+ * <p>The ORC file path and schema is passed during {@link OrcTableSource} construction. configuration is optional.
+ *
+ * <p>The OrcTableSource is used as shown in the example below.
+ *
+ * <pre>
+ * {@code
+ * String path = testInputURL.getPath();
+ * String schema = "struct<col1:boolean,col2:tinyint,col3:smallint,col4:int>"
+ * OrcTableSource orcSrc = new OrcTableSource(path, schema);
+ * tEnv.registerTableSource("orcTable", orcSrc);
+ * Table res = tableEnv.sql("SELECT * FROM orcTable");
+ * }
+ * </pre>
+ */
+public class OrcTableSource implements BatchTableSource<Row>, ProjectableTableSource<Row> {
+
+	private String path;
+	private TypeDescription orcSchema;
+	private RowTypeInfo typeInfo;
+	private Configuration orcConfig;
+	private int[] fieldMapping;
+
+	/**
+	 * The ORC file path and schema.
+	 *
+	 * @param path      the path of orc file
+	 * @param orcSchema schema of orc file
+	 */
+	public OrcTableSource(String path, String orcSchema) {
```

Comment:
Creating a new instance of `Configuration` is expensive. I would force the users to pass it in.